File: server/typescript/packages/codegen-ts/src/templates/fr010-field-mapping.ts
// server/typescript/packages/codegen-ts/src/templates/fr010-field-mapping.ts
//
// Shared field-kind mapping for the FR-010 codegen emitters (the output-format-spec
// emitter et al.). Maps a metadata field subtype onto the render engine's
// FieldKind member, the idiomatic nullable TS type used by the extract mirror interface,
// and the ExtractMap accessor that reads it from the forgiving outcome map.
//
// Mirrors the C# Fr010FieldMapping (adapted to TS syntax + the `| null` nullable mirror).
// Bounded scope (parity with Java/Kotlin/C#): scalar / enum / scalar-array. Nested object +
// array-of-enum are deferred.

import {
  type MetaData,
  TYPE_FIELD,
  FIELD_SUBTYPE_STRING,
  FIELD_SUBTYPE_UUID,
  FIELD_SUBTYPE_DATE,
  FIELD_SUBTYPE_TIME,
  FIELD_SUBTYPE_TIMESTAMP,
  FIELD_SUBTYPE_INT,
  FIELD_SUBTYPE_LONG,
  FIELD_SUBTYPE_CURRENCY,
  FIELD_SUBTYPE_DOUBLE,
  FIELD_SUBTYPE_FLOAT,
  FIELD_SUBTYPE_DECIMAL,
  FIELD_SUBTYPE_BOOLEAN,
  FIELD_SUBTYPE_ENUM,
  FIELD_SUBTYPE_OBJECT,
  FIELD_ATTR_REQUIRED,
  FIELD_ATTR_VALUES,
  FIELD_ATTR_COERCE_DEFAULT,
  FIELD_ATTR_DEFAULT,
  FIELD_ATTR_NORMALIZE,
  FIELD_ATTR_XML_TEXT,
  NORMALIZE_DEFAULT,
  type NormalizeMode,
} from "@metaobjectsdev/metadata";

/** The render-engine FieldKind member name for a scalar field subtype, or null if non-scalar. */
export function scalarKind(subType: string): string | null {
  switch (subType) {
    case FIELD_SUBTYPE_STRING:
    case FIELD_SUBTYPE_UUID:
    case FIELD_SUBTYPE_DATE:
    case FIELD_SUBTYPE_TIME:
    case FIELD_SUBTYPE_TIMESTAMP:
    // field.decimal is a precision-exact decimal STRING on the wire (not a
    // float64): extract/output map it as a string scalar so digits survive a
    // round-trip — matching the generated TS `string` representation.
    case FIELD_SUBTYPE_DECIMAL:
      return "STRING";
    case FIELD_SUBTYPE_INT:
      return "INT";
    case FIELD_SUBTYPE_LONG:
    case FIELD_SUBTYPE_CURRENCY:
      return "LONG";
    case FIELD_SUBTYPE_DOUBLE:
    case FIELD_SUBTYPE_FLOAT:
      return "DOUBLE";
    case FIELD_SUBTYPE_BOOLEAN:
      return "BOOLEAN";
    default:
      return null;
  }
}

/** The field children of a payload value-object, in declaration order. */
export function fields(vo: MetaData): MetaData[] {
  return vo.children().filter((c) => c.type === TYPE_FIELD);
}

/** isArray is a native (reserved) property on MetaData, not an attr. */
export function isArray(field: MetaData): boolean {
  return field.isArray === true;
}

/** True iff the field's @required is explicitly true (or the string "true"). */
export function isRequired(field: MetaData): boolean {
  const v = field.ownAttr(FIELD_ATTR_REQUIRED);
  if (v === true) return true;
  return typeof v === "string" && v.toLowerCase() === "true";
}

/** True iff the field's @xmlText is explicitly true (the XML text-content extract marker). */
export function xmlText(field: MetaData): boolean {
  const v = field.ownAttr(FIELD_ATTR_XML_TEXT);
  if (v === true) return true;
  return typeof v === "string" && v.toLowerCase() === "true";
}

/** The string members of an enum field's @values attr (empty when absent). */
export function enumValues(field: MetaData): string[] {
  const v = field.ownAttr(FIELD_ATTR_VALUES);
  if (Array.isArray(v)) return v.map((e) => String(e));
  return [];
}

/**
 * FR-011: the field's `@coerceDefault` member symbol (present-but-uncoercible enum fallback),
 * or null when absent. Read own-attr only — `@coerceDefault` is concrete, never inherited.
 */
export function coerceDefault(field: MetaData): string | null {
  const v = field.ownAttr(FIELD_ATTR_COERCE_DEFAULT);
  return typeof v === "string" && v.length > 0 ? v : null;
}

/**
 * FR-011: the field's `@default` member symbol (absent-fill enum value), or null when absent.
 */
export function defaultValue(field: MetaData): string | null {
  const v = field.ownAttr(FIELD_ATTR_DEFAULT);
  return typeof v === "string" && v.length > 0 ? v : null;
}

/**
 * FR-011: resolve the enum normalization mode for a field — field-level `@normalize`,
 * else the owning `object.value`'s `@normalize` (the per-object default), else the global
 * `NORMALIZE_DEFAULT` ("strip"). `ownerObject` may be null (resolution then skips the
 * object tier). Mirrors the cross-port field → object → global resolution.
 */
export function resolveNormalize(field: MetaData, ownerObject: MetaData | null): NormalizeMode {
  const fieldMode = normalizeAttrOf(field);
  if (fieldMode != null) return fieldMode;
  const objMode = ownerObject == null ? null : normalizeAttrOf(ownerObject);
  if (objMode != null) return objMode;
  return NORMALIZE_DEFAULT;
}

/** The `@normalize` attr of a node as a NormalizeMode, or null when absent. */
function normalizeAttrOf(node: MetaData): NormalizeMode | null {
  const v = node.ownAttr(FIELD_ATTR_NORMALIZE);
  return typeof v === "string" && v.length > 0 ? (v as NormalizeMode) : null;
}

/** The nullable TS type for a field in the extract mirror interface. */
export function mirrorType(field: MetaData): string {
  // Nested object (single or array): the self-contained path defers to null (typed unknown);
  // the runtime-delegating path overrides this with the nested mirror type. Checked BEFORE the
  // generic isArray branch so an array-of-objects is NOT mistyped as a string array.
  if (field.subType === FIELD_SUBTYPE_OBJECT) return "unknown"; // nested deferred
  // Matches asStringList's `(string | null)[] | null` return — a extracted array
  // can contain null elements where individual items were lost.
  if (isArray(field)) return "(string | null)[] | null";
  if (field.subType === FIELD_SUBTYPE_ENUM) return "string | null"; // enum is string-backed
  switch (scalarKind(field.subType)) {
    case "INT":
    case "LONG":
    case "DOUBLE":
      return "number | null";
    case "BOOLEAN":
      return "boolean | null";
    default:
      return "string | null";
  }
}

/**
 * The ExtractMap.as* helper name that reads this field from the forgiving map, or null
 * for a nested object (which emits a null literal, not a helper call). Single source of
 * truth for the per-field dispatch — both extractMapCall and extractMapHelpersUsed use it.
 */
function extractMapHelper(field: MetaData): string | null {
  // Nested object (single or array) → null literal in the self-contained path (no helper).
  // Checked BEFORE isArray so an array-of-objects is NOT read via asStringList (which would not
  // type-check against the nested mirror's `(NestedExtracted | null)[]`). Mirrors the Java fix.
  if (field.subType === FIELD_SUBTYPE_OBJECT) return null;
  if (isArray(field)) return "asStringList";
  if (field.subType === FIELD_SUBTYPE_ENUM) return "asString";
  switch (scalarKind(field.subType)) {
    case "INT":
      return "asInt";
    case "LONG":
      return "asLong";
    case "DOUBLE":
      return "asDouble";
    case "BOOLEAN":
      return "asBool";
    default:
      return "asString";
  }
}

/** The ExtractMap.as* helper name + call that reads this field from the forgiving map `d`. */
export function extractMapCall(field: MetaData): string {
  const helper = extractMapHelper(field);
  if (helper === null) return "null /* FR-010: nested extract deferred */";
  return `${helper}(d, ${jsonStringLiteral(field.name)})`;
}

/** Distinct ExtractMap helper names used across a value-object's fields (for the import). */
export function extractMapHelpersUsed(vo: MetaData): string[] {
  const used = new Set<string>();
  for (const f of fields(vo)) {
    const helper = extractMapHelper(f);
    if (helper !== null) used.add(helper);
  }
  return [...used].sort();
}

/** A TS double-quoted string literal for `value`, with the load-bearing chars escaped. */
export function jsonStringLiteral(value: string): string {
  let out = '"';
  for (const ch of value) {
    switch (ch) {
      case "\\":
        out += "\\\\";
        break;
      case '"':
        out += '\\"';
        break;
      case "\t":
        out += "\\t";
        break;
      case "\n":
        out += "\\n";
        break;
      case "\r":
        out += "\\r";
        break;
      default:
        out += ch;
    }
  }
  return out + '"';
}

/** A TS array literal `["a", "b"]` for the given members. */
export function stringArrayLiteral(values: readonly string[]): string {
  return "[" + values.map((v) => jsonStringLiteral(v)).join(", ") + "]";
}

/**
 * A TS object literal `{ "k": "v", … }` for a properties-shaped attr (e.g. @enumAlias /
 * @enumDoc), or "null" when absent/empty. Null values are dropped; keys are sorted
 * ordinally for deterministic output (matches the canonical-serializer properties sort).
 */
export function propertiesMapLiteral(attr: unknown): string {
  if (attr == null || typeof attr !== "object" || Array.isArray(attr)) return "null";
  const d = attr as Record<string, unknown>;
  const entries = Object.keys(d)
    .filter((k) => d[k] != null)
    .sort()
    .map((k) => `${jsonStringLiteral(k)}: ${jsonStringLiteral(String(d[k]))}`);
  if (entries.length === 0) return "null";
  return "{ " + entries.join(", ") + " }";
}

File: server/csharp/MetaObjects.Codegen/Generators/Fr010FieldMapping.cs
// Shared field-kind mapping for the FR-010 codegen emitters (RecoverSchemaEmitter +
// OutputFormatSpecEmitter). Maps a metadata field subtype onto the render engine's
// FieldKind, the idiomatic nullable C# type used by the recover mirror record, and the
// RecoverMap accessor that reads it from the forgiving outcome map.
//
// Mirrors the Java SpringTypeMapper / RecoverSchemaEmitter instanceof order and the C#
// PayloadCodegen scalar map. Bounded scope (parity with Java/Kotlin): scalar / enum /
// scalar-array. Nested object + array-of-enum are deferred (see KNOWN_GAPS).

using System.Collections.Generic;
using System.Linq;
using MetaObjects.Meta;
using static MetaObjects.Shared.BaseTypes;
using static MetaObjects.Shared.Structural;
using static MetaObjects.Core.Field.FieldConstants;

namespace MetaObjects.Codegen.Generators;

internal static class Fr010FieldMapping
{
    /// <summary>The field children of a payload value-object, in declaration order.</summary>
    public static IEnumerable<MetaData> Fields(MetaData vo) =>
        vo.Children().Where(c => c.Type == TYPE_FIELD);

    /// <summary>The string members of an enum field's <c>@values</c> attr (empty when absent).</summary>
    public static IReadOnlyList<string> EnumValues(MetaData field) =>
        field.OwnAttr(FIELD_ATTR_VALUES) switch
        {
            IReadOnlyList<string> ss => ss,
            IReadOnlyList<object?> os => os.Select(o => o?.ToString() ?? "").ToList(),
            _ => new List<string>(),
        };

    /// <summary>
    /// Emit a C# <c>new Dictionary&lt;string, string&gt; { … }</c> literal for a properties-shaped
    /// attr (e.g. <c>@enumDoc</c> / <c>@enumAlias</c>), or <c>"null"</c> when absent/empty.
    /// Null values are dropped; keys are sorted ordinally for deterministic output (matches the
    /// canonical-serializer properties sort).
    /// </summary>
    public static string PropertiesMapLiteral(object? attr)
    {
        if (attr is not IReadOnlyDictionary<string, object?> d) return "null";
        var entries = d.Where(kv => kv.Value is not null)
            .OrderBy(kv => kv.Key, System.StringComparer.Ordinal)
            .Select(kv => $"[\"{CSharpStringLiteral(kv.Key)}\"] = \"{CSharpStringLiteral(kv.Value!.ToString()!)}\"")
            .ToList();
        if (entries.Count == 0) return "null";
        return "new Dictionary<string, string> { " + string.Join(", ", entries) + " }";
    }

    /// <summary>Emit a C# <c>new[] { "a", "b" }</c> string-array literal for the given members.</summary>
    public static string StringArrayLiteral(IEnumerable<string> values) =>
        "new[] { " + string.Join(", ", values.Select(v => $"\"{v}\"")) + " }";

    /// <summary>
    /// FR-011: the field's <c>@coerceDefault</c> member symbol (present-but-uncoercible enum fallback),
    /// or null when absent. Own-attr only — <c>@coerceDefault</c> is concrete, never inherited.
    /// </summary>
    public static string? CoerceDefault(MetaData field) =>
        field.OwnAttr(FIELD_ATTR_COERCE_DEFAULT) is string s && s.Length > 0 ? s : null;

    /// <summary>FR-011: the field's <c>@default</c> member symbol (absent-fill enum value), or null when absent.</summary>
    public static string? DefaultValue(MetaData field) =>
        field.OwnAttr(FIELD_ATTR_DEFAULT) is string s && s.Length > 0 ? s : null;

    /// <summary>
    /// FR-011: resolve the enum normalization mode for a field — field-level <c>@normalize</c>, else
    /// the owning <c>object.value</c>'s <c>@normalize</c> (the per-object default), else the global
    /// <see cref="FieldConstants.NORMALIZE_DEFAULT"/> ("strip"). <paramref name="ownerObject"/> may be
    /// null (resolution then skips the object tier). Mirrors the TS resolveNormalize.
    /// </summary>
    public static string ResolveNormalize(MetaData field, MetaData? ownerObject)
    {
        string? fieldMode = NormalizeAttrOf(field);
        if (fieldMode != null) return fieldMode;
        string? objMode = ownerObject == null ? null : NormalizeAttrOf(ownerObject);
        if (objMode != null) return objMode;
        return NORMALIZE_DEFAULT;
    }

    /// <summary>The <c>@normalize</c> attr of a node as a mode string, or null when absent.</summary>
    private static string? NormalizeAttrOf(MetaData node) =>
        node.OwnAttr(FIELD_ATTR_NORMALIZE) is string s && s.Length > 0 ? s : null;

    /// <summary>Map a <c>@normalize</c> mode string onto the render-engine <c>NormalizeMode</c> member name.</summary>
    public static string NormalizeModeMember(string mode) => mode switch
    {
        "none"     => "NormalizeMode.None",
        "collapse" => "NormalizeMode.Collapse",
        _          => "NormalizeMode.Strip",
    };

    /// <summary>
    /// The nullable C# type for a SINGLE scalar field subtype in a recover mirror
    /// (e.g. <c>int?</c> / <c>long?</c> / <c>double?</c> / <c>bool?</c> / <c>string?</c>). Enum is
    /// string-backed (<c>string?</c>). Shared by the self-contained mirror, the nested-aware
    /// (delegating) mirror, and the scalar-ARRAY element type so the three stay in lock-step with
    /// PayloadCodegen's strict scalar map.
    /// </summary>
    public static string ScalarMirrorType(string subType) => subType switch
    {
        FIELD_SUBTYPE_ENUM => "string?",
        _ => ScalarKind(subType) switch
        {
            "Int" => "int?",
            "Long" => "long?",
            "Double" => "double?",
            "Decimal" => "decimal?",
            "Boolean" => "bool?",
            _ => "string?",
        },
    };

    /// <summary>
    /// The nullable C# element type for a scalar ARRAY in a recover mirror. Kept in lock-step with
    /// the array reader (<see cref="RecoverMapCall"/>'s <c>As*List</c> switch): only int/long/double/bool
    /// have kind-typed list readers; everything else (enum, decimal, date/time, string) is string-backed
    /// via <c>AsStringList</c>. Deliberately distinct from <see cref="ScalarMirrorType"/> (single scalars),
    /// where a single decimal is <c>decimal?</c> via <c>AsDecimal</c> — there is no <c>AsDecimalList</c>
    /// (decimal is single-only, SP-A), so a decimal array degrades to string here rather than emitting
    /// a mirror/reader type mismatch.
    /// </summary>
    public static string ScalarArrayElementType(string subType) => ScalarKind(subType) switch
    {
        "Int" => "int?",
        "Long" => "long?",
        "Double" => "double?",
        "Boolean" => "bool?",
        _ => "string?",
    };

    /// <summary>The render-engine <c>FieldKind</c> member name for a scalar field subtype, or null if non-scalar.</summary>
    public static string? ScalarKind(string subType) => subType switch
    {
        FIELD_SUBTYPE_STRING or FIELD_SUBTYPE_CLASS or
        FIELD_SUBTYPE_DATE or FIELD_SUBTYPE_TIME or FIELD_SUBTYPE_TIMESTAMP => "String",
        FIELD_SUBTYPE_INT or FIELD_SUBTYPE_SHORT or FIELD_SUBTYPE_BYTE => "Int",
        FIELD_SUBTYPE_LONG or FIELD_SUBTYPE_CURRENCY => "Long",
        FIELD_SUBTYPE_DOUBLE or FIELD_SUBTYPE_FLOAT => "Double",
        FIELD_SUBTYPE_DECIMAL => "Decimal",
        FIELD_SUBTYPE_BOOLEAN => "Boolean",
        _ => null,
    };

    /// <summary>The nullable C# type for a field in the SELF-CONTAINED recover mirror record.</summary>
    public static string MirrorType(MetaData field)
    {
        // Object BEFORE array (the object-before-isArray fix): a nested-object field — whether
        // single OR array — is deferred to null in the self-contained path, so it must NOT be
        // typed as a string list (that would force RecoverMapCall down the AsStringList branch
        // and mismatch the nested-aware mirror the delegating path shares). The delegating path
        // overrides this with nested-mirror typing (RecoverDelegateEmitter.NestedMirrorRecords).
        if (field.SubType == FIELD_SUBTYPE_OBJECT) return "object?"; // nested deferred (self-contained)
        // Scalar ARRAY: kind-type the element so the mirror list matches what the kind-typed
        // RecoverMap.As*List reader actually produces (int?/long?/double?/bool?/string?). Nullable
        // element: a recovered array can carry null where individual items were lost. NOTE: this
        // uses ScalarArrayElementType (NOT ScalarMirrorType) — decimal/date/enum arrays are
        // string-backed here because there is no AsDecimalList reader (decimal is single-only,
        // SP-A); a single decimal stays decimal? via ScalarMirrorType below.
        if (IsArray(field))
            return $"global::System.Collections.Generic.IReadOnlyList<{ScalarArrayElementType(field.SubType)}>?";
        return ScalarMirrorType(field.SubType);
    }

    /// <summary>The <c>RecoverMap.As*</c> call that reads this field from the forgiving map <c>d</c>.</summary>
    public static string RecoverMapCall(MetaData field)
    {
        string name = field.Name;
        // Object BEFORE array (object-before-isArray): a nested object/array-of-objects defers to
        // null on the self-contained path (it cannot map a List<NestedRecovered> from a flat map).
        if (field.SubType == FIELD_SUBTYPE_OBJECT) return "null /* FR-010: nested recover deferred (self-contained) */";
        // Scalar ARRAY: read via the kind-typed RecoverMap.As*List so the produced element type
        // matches the kind-typed mirror list (an int[] reads via AsIntList, etc.). Enum arrays stay
        // string-backed (AsStringList) — they are string-typed in the mirror.
        if (IsArray(field) && field.SubType != FIELD_SUBTYPE_ENUM)
            return ScalarKind(field.SubType) switch
            {
                "Int" => $"RecoverMap.AsIntList(d, \"{name}\")",
                "Long" => $"RecoverMap.AsLongList(d, \"{name}\")",
                "Double" => $"RecoverMap.AsDoubleList(d, \"{name}\")",
                "Boolean" => $"RecoverMap.AsBoolList(d, \"{name}\")",
                _ => $"RecoverMap.AsStringList(d, \"{name}\")",
            };
        if (IsArray(field)) return $"RecoverMap.AsStringList(d, \"{name}\")";
        if (field.SubType == FIELD_SUBTYPE_ENUM) return $"RecoverMap.AsString(d, \"{name}\")";
        return ScalarKind(field.SubType) switch
        {
            "Int" => $"RecoverMap.AsInt(d, \"{name}\")",
            "Long" => $"RecoverMap.AsLong(d, \"{name}\")",
            "Double" => $"RecoverMap.AsDouble(d, \"{name}\")",
            "Decimal" => $"RecoverMap.AsDecimal(d, \"{name}\")",
            "Boolean" => $"RecoverMap.AsBool(d, \"{name}\")",
            _ => $"RecoverMap.AsString(d, \"{name}\")",
        };
    }

    public static bool IsArray(MetaData field) =>
        field.OwnAttr(RESERVED_KEY_IS_ARRAY) is true || field.IsArray;

    public static bool IsRequired(MetaData field) =>
        field.OwnAttr(FIELD_ATTR_REQUIRED) is true ||
        (field.OwnAttr(FIELD_ATTR_REQUIRED) is string s && s.Equals("true", StringComparison.OrdinalIgnoreCase));

    /// <summary>Escape a value for embedding inside a C# double-quoted string literal.</summary>
    public static string CSharpStringLiteral(string value)
    {
        var sb = new System.Text.StringBuilder(value.Length + 4);
        foreach (char c in value)
        {
            switch (c)
            {
                case '\\': sb.Append("\\\\"); break;
                case '"': sb.Append("\\\""); break;
                case '\t': sb.Append("\\t"); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                default: sb.Append(c); break;
            }
        }
        return sb.ToString();
    }
}
